File: Files/(Plug-Ins)/Filters Plugs/Depth/DepthController.m
//
//  DepthController.m
//  PPMacho
//
//  Created by C.W. Betts on 9/10/13.
//
//

#import "DepthController.h"
#include <PlayerPROCore/PlayerPROCore.h>
#include <PlayerPROCore/PPPlug.h>

@interface DepthController ()

@end

@implementation DepthController

- (id)initWithWindow:(NSWindow *)window
{
    self = [super initWithWindow:window];
    if (self) {
		isMultipleIstanceSafe = YES;

		dispatch_block_t tmp = ^{
			SInt32			i, temp, Inc = self.depthAmmount;
			char			*Sample8Ptr = (char*) theData->data;
			short			*Sample16Ptr = ( short*) theData->data;
			
			if( Inc == 0) Inc = 1;
			switch( theData->amp)
			{
				case 8:
					Sample8Ptr += selectionStart;
					
					for( i = 0; i < selectionEnd - selectionStart; i++)
					{
						temp = *Sample8Ptr;
						
						temp *= Inc;
						temp /= 256;
						
						temp *= 256;
						temp /= Inc;
						
						*Sample8Ptr = temp;
						
						if( stereoMode)
						{
							Sample8Ptr++;
							i++;
						}
						
						Sample8Ptr++;
					}
					break;
					
				case 16:
					Sample16Ptr += selectionStart/2;						// Div 2, because it's in bytes !!!
					
					for( i = 0; i < (selectionEnd - selectionStart)/2; i++)	// Div 2, because it's in bytes !!!
					{
						temp = *Sample16Ptr;
						
						temp *= Inc;
						temp /= 0x0000FFFF;
						
						temp *= 0x0000FFFF;
						temp /= Inc;
						
						*Sample16Ptr = temp;
						
						if( stereoMode)
						{
							Sample16Ptr++;
							i++;
						}
						
						Sample16Ptr++;
					}
					break;
			}


		};
		
		self.plugBlock = tmp;
    }
    
    return self;
}

- (void)windowDidLoad
{
    [super windowDidLoad];
    
    // Implement this method to handle any initialization after your window controller's window has been loaded from its nib file.
}

@end

static OSErr mainDepth(void						*unused,
					   sData					*theData,
					   long					SelectionStart,
					   long					SelectionEnd,
					   PPInfoPlug				*thePPInfoPlug,
					   short					StereoMode)				// StereoMode = 0 apply on all channels, = 1 apply on current channel
{
	DepthController *controller = [[DepthController alloc] initWithWindowNibName:@"DepthController" infoPlug:thePPInfoPlug];
	controller.depthAmmount = 8;
	controller.theData = theData;
	controller.selectionStart = SelectionStart;
	controller.selectionEnd = SelectionEnd;
	controller.stereoMode = StereoMode ? YES : NO;
	
	return [controller runAsSheet];
}

// D7F6D8C0-FC86-48E4-A2B8-61BB681DE071
#define PLUGUUID CFUUIDGetConstantUUIDWithBytes(kCFAllocatorSystemDefault, 0xD7, 0xF6, 0xD8, 0xC0, 0xFC, 0x86, 0x48, 0xE4, 0xA2, 0xB8, 0x61, 0xBB, 0x68, 0x1D, 0xE0, 0x71)

#define PLUGMAIN mainDepth
#define PLUGINFACTORY DepthFactory

#include "CFPlugin-FilterBridge.c"

File: PlayerPRO 6/Plug-ins/Filters Plugs/Depth/DepthController.swift
//
//  DepthController.swift
//  PPMacho
//
//  Created by C.W. Betts on 9/11/14.
//
//

import Cocoa
import PlayerPROKit

class DepthController: NSWindowController {
	dynamic var depthAmmount: Int32 = 0
	var selectionRange = NSRange(location: 0, length: 1)
	var stereoMode = false
	var currentBlock: PPPlugErrorBlock!
	var theData: PPSampleObject!
	
	var parentWindow: NSWindow!
	
    override func windowDidLoad() {
        super.windowDidLoad()
    
        // Implement this method to handle any initialization after your window controller's window has been loaded from its nib file.
    }

	@IBAction func okay(_ sender: AnyObject!) {
		let ourData = NSMutableData(data: theData.data)
		let rawData = UnsafeMutableRawPointer(ourData.mutableBytes)
		var Sample8Ptr = rawData.assumingMemoryBound(to: Int8.self)
		var Sample16Ptr = rawData.assumingMemoryBound(to: Int16.self)
		var Inc = depthAmmount
		if Inc == 0 {
			Inc = 1
		}
		
		switch theData.amplitude {
		case 8:
			Sample8Ptr = Sample8Ptr.advanced(by: selectionRange.location)
			var i = 0
			while i < selectionRange.length {
				var temp = Int32(Sample8Ptr.pointee)
				
				temp *= Inc;
				temp /= Int32(UInt8.max)
				
				temp *= Int32(UInt8.max)
				temp /= Inc;
				
				Sample8Ptr.pointee = Int8(temp)
				
				if stereoMode {
					Sample8Ptr = Sample8Ptr.successor()
					i += 1
				}
				
				Sample8Ptr = Sample8Ptr.successor()
				i += 1
			}
		case 16:
			Sample16Ptr = Sample16Ptr.advanced(by: selectionRange.location)
			var i = 0
			while i < selectionRange.length / 2 {	// Div 2, because it's in bytes !!!
				var temp = Int32(Sample16Ptr.pointee)
				
				temp *= Inc;
				temp /= Int32(UInt16.max)
				
				temp *= Int32(UInt16.max)
				temp /= Inc
				
				Sample16Ptr.pointee = Int16(temp)
				
				if stereoMode {
					Sample16Ptr = Sample16Ptr.successor()
					i += 1
				}
				
				Sample16Ptr = Sample16Ptr.successor()
				i += 1
			}
			
		default:
			NSBeep()
			parentWindow.endSheet(window!)
			currentBlock(.parametersErr)
			return
			
		}

		theData.data = ourData as Data
		parentWindow.endSheet(window!)
		currentBlock(.noErr);
	}
	
	@IBAction func cancel(_ sender: AnyObject!) {
		parentWindow.endSheet(window!)
		currentBlock(.userCanceledErr);
	}

}
